src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,296 @@
+                let mut col_map = HashMap::new();
+                let mut input_proj = vec![];
+                for i in 0..input_arity {
+                    if needed.contains(&i) {
+                        col_map.insert(i, input_proj.len());
+                        input_proj.push(i);
+                    }
+                }
+
+                for expr in exprs.iter_mut() {
+                    expr.visit_mut(&mut |e| {
+                        if let ScalarExpr::Column(c) = e {
+                            *c = *col_map.get(c).unwrap();
+                        }
+                    });
+                }
+                for c in projection.iter_mut() {
+                    if *c < input_arity {
+                        *c = *col_map.get(c).unwrap();
+                    } else {
+                        *c = *c - input_arity + input_proj.len();
+                    }
+                }
+                self.action(input, input_proj);
+                *relation = input
+                    .take_dangerous()
+                    .flat_map(func.clone(), exprs.clone())
+                    .project(projection);
+            }
+            RelationExpr::Map { input, scalars } => {
+                let input_arity = input.arity();
+                let mut needed: HashSet<_> = projection.iter().cloned().collect();
+                // First, prune any elements of scalars that are ignored by the projection.
+                // Compute the transitive closure of the needed columns.
+                let needed = loop {
+                    let mut new = needed.clone();
+                    for c in needed.iter() {
+                        if *c >= input_arity {
+                            new.extend(scalars[*c - input_arity].support());
+                        }
+                    }
+                    if new == needed {
+                        break new;
+                    }
+                    needed = new;
+                };
+                // First, prune away unneeded columns from the input.
+                let mut col_map = HashMap::new();
+                let mut input_proj = vec![];
+                for i in 0..input_arity {
+                    if needed.contains(&i) {
+                        col_map.insert(i, input_proj.len());
+                        input_proj.push(i);
+                    }
+                }
+
+                let mut new_scalars = vec![];
+                for (idx, expr) in scalars.iter_mut().enumerate() {
+                    if needed.contains(&(idx + input_arity)) {
+                        col_map.insert(idx + input_arity, input_proj.len() + new_scalars.len());
+                        expr.visit_mut(&mut |e| {
+                            if let ScalarExpr::Column(c) = e {
+                                *c = *col_map.get(c).unwrap();
+                            }
+                        });
+                        new_scalars.push(expr.clone());
+                    }
+                }
```

Comment:
I think maybe as above, if we start with the list of output columns that are "needed", potentially transitively, we can use that (sorted) to drive this whole process. Binary search enables the rewrites, and you may avoid the `col_map` entirely. Also `input_proj` is just that collection but `retain()`d to keep values less than `input_arity`.

All just stylistic, but I figure if these things can turn 20 lines in to 3-5 over and over, that makes it all much better.